`cannot take the address of an rvalue` when pass return value of function as interface.

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 18.04.5 LTS (WSL 2)
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 3800X 8-Core Processor
CC version: clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)

getwd: /home/zakuro/src/github.com/zakuro9715/v
vmodules: /home/zakuro/.vmodules
vroot: /home/zakuro/src/github.com/zakuro9715/v
vexe: /home/zakuro/src/github.com/zakuro9715/v/v
vexe mtime: 2021-05-07 09:48:35
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 13ea92a
env VFLAGS: "-cc clang"

Git version: git version 2.31.1
Git vroot status: weekly.2021.18-15-g13ea92a7-dirty
.git/config present: true
thirdparty/tcc status: master 17d18c13
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg interface_test.v`
```v
struct Cat {
mut:
	breed string
}

fn new_cat(breed string) Cat {
	return Cat{breed}
}

fn (c &Cat) speak(s string) {
	if c.breed != '' {
		assert c.breed == 'Persian'
	}
	assert s == 'Hi !'
	println('meow')
}

fn perform_speak(a Animal) {
	a.speak('Hi !')
}

fn test_perform_speak() {
	perform_speak(new_cat('Persian'))
}

interface Animal {
	speak(s string)
}

```

**What did you expect to see?**

success

**What did you see instead?**
```
==================
/tmp/v/interface_test.18366391640576003529.tmp.c:13825:60: error: cannot take the address of an rvalue of type 'main__Cat' (aka 'struct main__Cat')
        main__perform_speak(I_main__Cat_to_Interface_main__Animal(&main__new_cat(_SLIT("Persian"))));
                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1686,6 +1686,25 @@ fn (mut g Gen) write_sumtype_casting_fn(got_ ast.Type, exp_ ast.Type) {
 	g.auto_fn_definitions << sb.str()
 }
 
+fn (mut g Gen) call_cfn_for_casting_expr(fname string, expr ast.Expr, exp_is_ptr bool, exp_styp string, got_is_ptr bool, got_styp string) {
+	mut rparen_n := 1
+	if exp_is_ptr {
+		g.write('HEAP($exp_styp, ')
+		rparen_n++
+	}
+	g.write('${fname}(')
+	if !got_is_ptr {
+		if expr.is_lvalue() {
+			g.write('&')
+		} else {
+			g.write('ADDR($got_styp, (')
+			rparen_n += 2
+		}
+	}
+	g.expr(expr)
+	g.write(')'.repeat(rparen_n))
+}
+
 // use instead of expr() when you need to cast to a different type
 fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_type ast.Type) {
 	got_type := g.table.mktyp(got_type_raw)
@@ -1705,18 +1724,9 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 		&& !expected_type.has_flag(.optional) {
 		got_styp := g.cc_type(got_type, true)
 		exp_styp := g.cc_type(expected_type, true)
-		if expected_is_ptr {
-			g.write('HEAP($exp_styp, ')
-		}
-		g.write('I_${got_styp}_to_Interface_${exp_styp}(')
-		if !got_is_ptr {
-			g.write('&')
-		}
-		g.expr(expr)
-		g.write(')')
-		if expected_is_ptr {
-			g.write(')')
-		}
+		fname := 'I_${got_styp}_to_Interface_$exp_styp'
+		g.call_cfn_for_casting_expr(fname, expr, expected_is_ptr, exp_styp, got_is_ptr,
+			got_styp)
 		return
 	}
 	// cast to sum type
@@ -1745,21 +1755,9 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 				g.expr(expr)
 			} else {
 				g.write_sumtype_casting_fn(got_type, expected_type)
-				if expected_is_ptr {
-					g.write('HEAP($exp_sym.cname, ')
-				}
-				g.write('${got_sym.cname}_to_sumtype_${exp_sym.cname}(')
-				if !got_is_ptr {
-					g.write('ADDR($got_styp, (')
-					g.expr(expr)
-					g.write(')))')
-				} else {
-					g.expr(expr)
-					g.write(')')
-				}
-				if expected_is_ptr {
-					g.write(')')
-				}
+				fname := '${got_sym.cname}_to_sumtype_$exp_sym.cname'
+				g.call_cfn_for_casting_expr(fname, expr, expected_is_ptr, exp_sym.cname,
+					got_is_ptr, got_styp)
 			}
 			return
 		}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1708,11 +1708,12 @@ fn (mut g Gen) call_cfn_for_casting_expr(fname string, expr ast.Expr, exp_is_ptr
 	}
 	g.write('${fname}(')
 	if !got_is_ptr {
-		if expr.is_lvalue() {
-			g.write('&')
-		} else {
+		if !expr.is_lvalue()
+			|| (expr is ast.Ident && is_simple_define_const((expr as ast.Ident).obj)) {
 			g.write('ADDR($got_styp, (')
 			rparen_n += 2
+		} else {
+			g.write('&')
 		}
 	}
 	g.expr(expr)
@@ -5028,11 +5029,6 @@ fn (mut g Gen) const_decl(node ast.ConstDecl) {
 		*/
 		field_expr := field.expr
 		match field.expr {
-			ast.CharLiteral, ast.FloatLiteral, ast.IntegerLiteral {
-				// "Simple" expressions are not going to need multiple statements,
-				// only the ones which are inited later, so it's safe to use expr_string
-				g.const_decl_simple_define(name, g.expr_string(field_expr))
-			}
 			ast.ArrayInit {
 				if field.expr.is_fixed {
 					styp := g.typ(field.expr.typ)
@@ -5062,10 +5058,26 @@ fn (mut g Gen) const_decl(node ast.ConstDecl) {
 				}
 			}
 			else {
-				g.const_decl_init_later(field.mod, name, field.expr, field.typ, false)
+				if is_simple_define_const(field) {
+					// "Simple" expressions are not going to need multiple statements,
+					// only the ones which are inited later, so it's safe to use expr_string
+					g.const_decl_simple_define(name, g.expr_string(field_expr))
+				} else {
+					g.const_decl_init_later(field.mod, name, field.expr, field.typ, false)
+				}
 			}
 		}
 	}
+}
+
+fn is_simple_define_const(obj ast.ScopeObject) bool {
+	if obj is ast.ConstField {
+		return match obj.expr {
+			ast.CharLiteral, ast.FloatLiteral, ast.IntegerLiteral { true }
+			else { false }
+		}
+	}
+	return false
 }
 
 fn (mut g Gen) const_decl_simple_define(name string, val string) {

--- a/vlib/v/tests/interface_test.v
+++ b/vlib/v/tests/interface_test.v
@@ -7,6 +7,10 @@ mut:
 	breed string
 }
 
+fn new_cat(breed string) Cat {
+	return Cat{breed}
+}
+
 fn (c &Cat) name() string {
 	if c.breed != '' {
 		assert c.breed == 'Persian'
@@ -101,10 +105,12 @@ fn test_perform_speak() {
 	perform_speak(Cat{
 		breed: 'Persian'
 	})
+	perform_speak(new_cat('Persian'))
 	perform_speak_on_ptr(cat)
 	perform_speak_on_ptr(Cat{
 		breed: 'Persian'
 	})
+	perform_speak_on_ptr(new_cat('Persian'))
 	handle_animals([dog, cat])
 	/*
 	f := Foo {

--- a/vlib/v/tests/sumtype_calls_test.v
+++ b/vlib/v/tests/sumtype_calls_test.v
@@ -56,9 +56,13 @@ fn test_sum_type_cast() {
 fn test_sum_types() {
 	b := parse_bool()
 	handle_expr(b)
+	handle_expr(parse_bool())
+
 	de := DeclExprA{}
 	handle_expr(de)
 	handle_decl_expr(de)
+	handle_expr(DeclExprA{})
+	handle_decl_expr(DeclExprA{})
 }
 
 /*

